ring/consul_client.go

```diff
@@ -64,25 +72,112 @@ var (
 	ErrNotFound = fmt.Errorf("Not found")
 )
 
-// Get and deserialise a JSON value from Consul.
-func (c *consulClient) Get(key string, factory InstanceFactory) error {
-	kvp, _, err := c.kv.Get(key, queryOptions)
+// ProtoSerDes is a SerDes for proto/snappy
+type ProtoSerDes struct {
+	Factory func() proto.Message
+}
+
+// Decode implements SerDes
+func (p ProtoSerDes) Decode(bytes []byte) (interface{}, error) {
+	out := p.Factory()
+	bytes, err := snappy.Decode(nil, bytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := proto.Unmarshal(bytes, out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Encode implements SerDes
+func (p ProtoSerDes) Encode(msg interface{}) ([]byte, error) {
+	bytes, err := proto.Marshal(msg.(proto.Message))
 	if err != nil {
-		return err
+		return nil, err
+	}
+	return snappy.Encode(nil, bytes), nil
+}
+
+// JSONSerDes is a SerDes for JSON
+type JSONSerDes struct {
+	Factory func() interface{}
+}
+
+// Decode implements SerDes
+func (j JSONSerDes) Decode(bytes []byte) (interface{}, error) {
+	out := j.Factory()
+	if err := json.Unmarshal(bytes, out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Encode implemenrs SerDes
+func (j JSONSerDes) Encode(msg interface{}) ([]byte, error) {
+	return json.Marshal(msg)
+}
+
+// DynamicSerDes is a SerDes that can read json and proto, and
+// that can serialise to either (selectively).
+// Once it fails to decode JSON, it will start decoding (and
+// writing) protos.
+type DynamicSerDes struct {
+	mtx      sync.Mutex
+	useProto bool
+	json     SerDes
+	proto    SerDes
+}
+
+// NewDynamicSerDes makes a new DynamicSerDes
+func NewDynamicSerDes(json, proto SerDes) *DynamicSerDes {
+	return &DynamicSerDes{
+		useProto: false,
+		json:     json,
+		proto:    proto,
 	}
-	if kvp == nil {
-		return ErrNotFound
+}
+
+// UseProto allow you to change the SerDes at runtime.
+func (d *DynamicSerDes) UseProto(useProto bool) {
+	log.Infof("Switching to proto serialization: %v", useProto)
+
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	d.useProto = useProto
+}
+
+// Decode implements SerDes
+func (d *DynamicSerDes) Decode(bytes []byte) (interface{}, error) {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+
+	out, err := d.json.Decode(bytes)
+	if err == nil {
+		return out, nil
 	}
-	out := factory()
-	if err := json.NewDecoder(bytes.NewReader(kvp.Value)).Decode(out); err != nil {
-		return err
+
+	if !d.useProto {
+		log.Infof("Error decoding json, switching to writing proto: %v", err)
+		d.useProto = true
 	}
-	return nil
+
+	return d.proto.Decode(bytes)
```

Comment:
The downside of this method is that it will mask genuine JSON errors. A more precise approach would be to figure out exactly _which_ error is returned when we try to decode protobufs as JSON (https://godoc.org/encoding/json has a bunch), and then type assert on that. 

I'm not entirely sure it's worth it, though.